packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+ * @param {*} redis
+ * @param {string} cacheKey
+ * @param {number} cacheTtl
+ * @param {() => Promise<T>} refreshFunction
+ * @returns {Promise<T>} possibly cached result
+ */
+async function cachedResult(log, redis, cacheKey, cacheTtl, refreshFunction) {
+  if (cacheTtl) {
+    try {
+      const json = await redis.get(cacheKey);
+      if (json) {
+        return JSON.parse(json);
+      }
+    } catch (err) {
+      log.error(
+        `subhub.cachedResult.${refreshFunction.name}.getCachedResponse.failed`,
+        { err }
+      );
+    }
+  }
+
+  const result = await refreshFunction();
+  if (cacheTtl) {
+    redis
+      .set(cacheKey, JSON.stringify(result), 'EX', cacheTtl)
+      .catch(err =>
+        log.error(
+          `subhub.cachedResult.${refreshFunction.name}.cacheResponse.failed`,
+          { err }
+        )
+      );
+  }
+  return result;
+}
+
+class StripeHelper {
+  /**
+   * Create a Stripe Helper with built-in caching.
+   *
+   * @param {object} log
+   * @param {object} config
+   */
+  constructor(log, config) {
+    this.log = log;
+    this.cacheTtlSeconds = config.subhub.plansCacheTtlSeconds;
+    const redis =
+      this.cacheTtlSeconds &&
+      require('../redis')(
+        {
+          ...config.redis,
+          ...config.redis.subhub,
+        },
+        log
+      );
+    this.plansCacheIsEnabled = this.cacheTtlSeconds && redis;
+
+    this.stripe = new stripe(config.subscriptions.stripeApiKey, {
+      maxNetworkRetries: 3,
+    });
+    this.redis = redis;
+  }
+
+  /**
+   * Fetch all product data and cache it if Redis is enabled.
+   *
+   * Use `allProducts` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevProduct[]>} All the products.
+   */
+  async fetchAllProducts() {
+    const products = [];
+    for await (const product of this.stripe.products.list()) {
+      products.push({
+        product_id: product.id,
+        product_name: product.name,
+        product_metadata: product.metadata,
+      });
+    }
+    return products;
+  }
+
+  /**
+   * Fetches all products from stripe and returns them.
+   *
+   * Uses Redis caching if configured.
+   *
+   * @returns {Promise<AbbrevProduct[]>} All the products.
+   */
+  async allProducts() {
+    return cachedResult(
+      this.log,
+      this.redis,
+      'listProducts',
+      this.cacheTtlSeconds,
+      () => this.fetchAllProducts()
+    );
+  }
+
+  /**
+   * Fetches all plans from stripe and returns them.
+   *
+   * Use `allPlans` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevPlan[]>} All the plans.
+   */
+  async fetchAllPlans() {
+    const plans = [];
+    for await (const item of this.stripe.plans.list({
+      active: true,
+      expand: ['data.product'],
+    })) {
+      // FIXME: Should probably error here if we can't set a product id/name.
+      let product_id, product_name, product_metadata;
+      if (item.product && typeof item.product !== 'string') {
```

Comment:
What if item.product IS a string? Should that be handled too?

Reply:
If item.product is a string, it means Stripe didn't expand the product like it was directed to. We'll probably want to error out to force a retry per the FIXME.